metrics.go

```diff
@@ -0,0 +1,55 @@
+package dht
+
+import (
+	"go.opencensus.io/stats"
+	"go.opencensus.io/stats/view"
+	"go.opencensus.io/tag"
+)
+
+// Copied from https://github.com/census-instrumentation/opencensus-go/blob/master/plugin/ocgrpc/stats_common.go
+var (
+	DefaultBytesDistribution        = view.Distribution(1024, 2048, 4096, 16384, 65536, 262144, 1048576, 4194304, 16777216, 67108864, 268435456, 1073741824, 4294967296)
+	DefaultMillisecondsDistribution = view.Distribution(0.01, 0.05, 0.1, 0.3, 0.6, 0.8, 1, 2, 3, 4, 5, 6, 8, 10, 13, 16, 20, 25, 30, 40, 50, 65, 80, 100, 130, 160, 200, 250, 300, 400, 500, 650, 800, 1000, 2000, 5000, 10000, 20000, 50000, 100000)
+	DefaultMessageCountDistribution = view.Distribution(1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096, 8192, 16384, 32768, 65536)
+)
+
+// Keys
+var (
+	// TODO: figure out some appropriate keys
+	// TODO: add a key for each query type to be able to tag every query made
+	// KeyProtocol, _     = tag.NewKey("libp2p_protocol")
+	// KeyPeerID, _       = tag.NewKey("libp2p_peer_id")
+	// KeyRemotePeerID, _ = tag.NewKey("libp2p_remote_peer_id")
+	KeyMessageType, _ = tag.NewKey("message_type")
+)
+
+// Metrics
+var (
+	ReceivedMessagesPerRPC = stats.Int64("libp2p.io/dht/kad/received_messages_per_rpc", "Total number of messages received per RPC", stats.UnitDimensionless)
+	ReceivedBytesPerRPC    = stats.Int64("libp2p.io/dht/kad/received_bytes_per_rpc", "Total received bytes per RPC", stats.UnitBytes)
+	LatencyPerRPC          = stats.Float64("libp2p.io/dht/kad/latency_per_rpc", "Latency per RPC", stats.UnitMilliseconds)
+)
+
+// Views
+var (
+	ReceivedMessagesPerRPCView = &view.View{
```

Comment:
I think these should be returned from a method on the node:
```
func (dht *IpfsDHT) GetMetrics() NodeMetrics { return dht.metrics }

type NodeMetrics struct {
    ReceiveMessagesPerRpcView view.View
    ReceivedBytes...
    Latency...
}

struct IpfsDHT {
    ...
    metrics NodeMetrics
}
```

Reply:
Private views would prevent users of the dht package from picking and choosing which metrics they want to aggregate, it makes it an all or nothing scenario which reduces one of the main benefits of opencensus. We could put the metrics stuff in a `stats` subpackage so the main `dht` package isn't littered with the extra exported vars.

Reply:
I think you've misunderstood. The stats can be localized to the instance, and exposed through a method. Mainly it tidies up the package namespace and provides partitioning of stats for nodes. It has no affect on the consumer, they can choose to register some of the views, or not, just the same.

Reply:
The stats are already localised by node based on the register. See the below image that has the metrics divided by type and instance (node).
![1552434347](https://user-images.githubusercontent.com/5924712/54247497-ed417380-4584-11e9-863f-9765de06f461.png)

So there is no need to embedded the metrics into `IpfsDHT`. We can create a utility function that returns all the views for ease of registering but that won't remove them from the package namespace. Either way the views have to be exported for consumption by the packages that register them and there is no real benefit to putting them in a struct.